src/p2p/behaviour.rs

```diff
@@ -419,16 +420,29 @@ impl<Types: IpfsTypes> Behaviour<Types> {
         self.swarm.disconnect(addr)
     }
 
+    // FIXME: it would be best if get_providers is called only in case the already connected
+    // peers don't have it
     pub fn want_block(&mut self, cid: Cid) {
-        //let hash = Multihash::from_bytes(cid.to_bytes()).unwrap();
-        //self.kademlia.get_providers(hash);
+        let key = cid.to_bytes();
+        self.kademlia.get_providers(key.into());
         self.bitswap.want_block(cid, 1);
     }
 
+    // FIXME: it would probably be best if this could return a SubscriptionFuture, so
+    // that the put_block operation truly finishes only when the block is already being
+    // provided; it is, however, pretty tricky in terms of internal communication between
+    // Ipfs and IpfsFuture objects - it would currently require some extra back-and-forth
     pub fn provide_block(&mut self, cid: Cid) {
```

Comment:
I am not 100% with you on this FIXME but no need to change it now. To elaborate:

"Not 100% with you" as in is that the saving blocks should be as decoupled as possible and providing should only happen for some subset of blocks: possibly only the pinned roots, or pinned roots and their subroots or anything more complex would need to be provided for. Where "subroots" would be the roots of the files in a pinned larger directory structure for the invidiual files.

Decoupling providing and put_block because providing can be quite expensive, and we might not want to do them concurrently at all. Additionally not sure if it's a good idea to plan to use start_providing unless we can come up with a really nice backend for the kademlia store. Otherwise we could just use a simpler and handle the timing ourselves with the "push" api.

Reply:
Fair enough :+1: